editor/editor_notes.cpp

```diff
@@ -0,0 +1,213 @@
+               uint32_t const uploadedNotesCount)
+{
+  auto constexpr kDigitsAfterComma = 7;
+  auto root = xml.append_child("notes");
+  root.append_attribute("uploadedNotesCount") = uploadedNotesCount;
+  for (auto const & note : notes)
+  {
+    auto node = root.append_child("note");
+
+    node.append_attribute("lat") = strings::to_string_dac(note.m_point.lat, kDigitsAfterComma).data();
+    node.append_attribute("lon") = strings::to_string_dac(note.m_point.lon, kDigitsAfterComma).data();
+    node.append_attribute("text") = note.m_note.data();
+  }
+}
+
+/// Not thread-safe, use only for initialization.
+bool Load(string const & fileName, list<editor::Note> & notes, uint32_t & uploadedNotesCount)
+{
+  string content;
+  try
+  {
+    auto const reader = GetPlatform().GetReader(fileName);
+    reader->ReadAsString(content);
+  }
+  catch (FileAbsentException const &)
+  {
+    // It's normal if no notes file is present.
+    return true;
+  }
+  catch (Reader::Exception const & e)
+  {
+    LOG(LERROR, ("Can't process file.", fileName, e.Msg()));
+    return false;
+  }
+
+  pugi::xml_document xml;
+  if (!xml.load_buffer(content.data(), content.size()))
+  {
+    LOG(LERROR, ("Can't load notes, XML is ill-formed.", content));
+    return false;
+  }
+
+  notes.clear();
+  if (!LoadFromXml(xml, notes, uploadedNotesCount))
+  {
+    LOG(LERROR, ("Can't load notes, file is ill-formed.", content));
+    return false;
+  }
+
+  return true;
+}
+
+/// Not thread-safe, use synchronization.
+bool Save(string const & fileName,
+          list<editor::Note> const & notes,
+          uint32_t const uploadedNotesCount)
+{
+  pugi::xml_document xml;
+  SaveToXml(notes, xml, uploadedNotesCount);
+  return my::WriteToTempAndRenameToFile(
+      fileName,
+      [&xml](string const & fileName)
+      {
+        return xml.save_file(fileName.data(), "  ");
+      });
+}
+}  // namespace
+
+namespace editor
+{
+shared_ptr<Notes> Notes::MakeNotes(string const & fileName, bool const fullPath)
+{
+  return shared_ptr<Notes>(
+      new Notes(
+          fullPath
+          ? fileName
+          : GetPlatform().WritablePathForFile(fileName)));
+}
+
+Notes::Notes(string const & fileName)
+    : m_fileName(fileName)
+{
+  Load(m_fileName, m_notes, m_uploadedNotesCount);
+}
+
+void Notes::CreateNote(m2::PointD const & point, string const & text)
+{
+  lock_guard<mutex> g(m_mu);
+  m_notes.emplace_back(MercatorBounds::ToLatLon(point), text);
+  Save(m_fileName, m_notes, m_uploadedNotesCount);
+}
+
+void Notes::Upload(osm::OsmOAuth const & auth)
+{
+  // Capture self to keep it from destruction until this thread is done.
+  auto const self = shared_from_this();
+
+  auto const doUpload = [self, auth]()
+  {
+    size_t size;
+
+    {
+      lock_guard<mutex> g(self->m_mu);
+      size = self->m_notes.size();
+    }
+
+    osm::ServerApi06 api(auth);
+    auto it = begin(self->m_notes);
+    for(size_t i = 0; i != size; ++i, ++it)
```

Comment:
Нет ненужной переменной.

> On Mar 16, 2016, at 11:17, mgsergio <notifications@github.com> wrote:
> 
> In editor/editor_notes.cpp <https://github.com/mapsme/omim/pull/2178#discussion_r56295609>:
> 
> > +{
> > +  // Capture self to keep it from destruction until this thread is done.
> > +  auto const self = shared_from_this();
> > +
> > +  auto const doUpload = [self, auth]()
> > +  {
> > +    size_t size;
> > +
> > +    {
> > +      lock_guard<mutex> g(self->m_mu);
> > +      size = self->m_notes.size();
> > +    }
> > +
> > +    osm::ServerApi06 api(auth);
> > +    auto it = begin(self->m_notes);
> > +    for(size_t i = 0; i != size; ++i, ++it)
> Какая разница?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub <https://github.com/mapsme/omim/pull/2178/files/43432b48627ff39f7762b4e5f1940fb3f9f8c59f#r56295609>